src/Microsoft.Extensions.Http/DefaultHttpClientFactory.cs

```diff
@@ -87,16 +144,154 @@ internal HttpMessageHandler CreateHandler(string name)
                 for (var i = 0; i < options.HttpMessageHandlerBuilderActions.Count; i++)
                 {
                     options.HttpMessageHandlerBuilderActions[i](b);
                 }
             }
         }
+
+        // Internal for tests
+        internal void ExpiryTimer_Tick(object state)
+        {
+            var active = (ActiveHandlerTrackingEntry)state;
+            active.DisposeTimer();
+
+            // The timer callback should be the only one removing from the active collection. If we can't find
+            // our entry in the collection, then this is a bug.
+            var removed = _activeHandlers.TryRemove(active.Name, out var found);
+            Debug.Assert(!removed);
+            Debug.Assert(object.ReferenceEquals(active, found));
+
+            // At this point the handler is no longer 'active' and will not be handed out to any new clients.
+            // However we haven't dropped our strong reference to the handler, so we can't yet determine if
+            // there are still any other outstanding references (we know there is at least one).
+            //
+            // We use a different state object to track expired handlers. This allows any other thread that acquired
+            // the 'active' entry to use it without safety problems.
+            var expired = new ExpiredHandlerTrackingEntry(active);
+            _expiredHandlers.Enqueue(expired);
+
+            StartCleanupTimer();
+        }
+
+        // Internal so it can be overridden in tests
+        internal virtual void StartCleanupTimer()
+        {
+            lock (_cleanupTimerLock)
+            {
+                if (_cleanupTimer == null)
+                {
+                    _cleanupTimer = new Timer(_cleanupCallback, null, DefaultCleanupInterval, Timeout.InfiniteTimeSpan);
+                }
+            }
+        }
+
+        // Internal for tests
+        internal void CleanupTimer_Tick(object state)
+        {
+            // Stop any pending timers, we'll restart the timer if there's anything left to process after cleanup.
+            //
+            // With the scheme we're using it's possible we could end up with some redundant cleanup operations.
+            // This is expected and fine.
+            // 
+            // An alternative would be to take a lock during the whole cleanup process. This isn't ideal because it
+            // would result in threads executing ExpiryTimer_Tick as they would need to block on cleanup to figure out
+            // whether we need to start the timer.
+            lock (_cleanupTimerLock)
+            {
+                _cleanupTimer.Dispose();
+                _cleanupTimer = null;
+            }
+
+            var stopwatch = ValueStopwatch.StartNew();
+
+            var initialCount = _expiredHandlers.Count;
+            Log.CleanupCycleStart(_logger, initialCount);
+
+            var disposedCount = 0;
+            for (var i = 0; i < initialCount; i++)
```

Comment:
There's actually a subtle issue here, we need to prevent cleanup cycles from running concurrently if the time the cycle takes is longer than the period. 

We really don't intend for multiple threads to do cleanup at once. I could make it supported, but it seems better to prevent it.

Will fix in the next iteration.